### PR TITLE
CDAP-13262 Expose metadata api to plugins

### PR DESCRIPTION
**Pull request by rohitsinha54**
## Summary
- Exposes Metadata API to plugins
- Metadata propagation / processing is the responsibility of the plugin developer
- Added a unit test where metadata operation JSON is passed as plugin config to a mock plugin for testing

## Tasks
- [x] Build: https://builds.cask.co/browse/CDAP-DUT6529-1
- [ ] Testing on multi-node cluster
- [ ] Test on multi-node cluster with spark engine

### CODE REVIEW

File: cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java [hunk 1]
```diff
@@ -113,6 +113,22 @@
    */
   public abstract MessagingContext getMessagingContext();
 
+  /**
+   * Returns a {@link MetadataReader} which can be used to read metadata. Currently the returned instance can
+   * only be used in the Spark driver process.
+   *
+   * @return A {@link MetadataReader}
+   */
+  public abstract MetadataReader getMetadataReader();
```
Comment (chtyim) [inline cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java]: Why only available for Java? Every spark API should be exposed through the Scala one as well.

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java [hunk 2]
```diff
@@ -50,6 +61,11 @@
 @Plugin(type = BatchSource.PLUGIN_TYPE)
 @Name("Mock")
 public class MockSource extends BatchSource<byte[], Row, StructuredRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MockSource.class);
+
+  public static final Gson GSON = new Gson();
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Why public??

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java [hunk 3]
```diff
@@ -175,6 +214,52 @@ private static PluginClass getPluginClass() {
...
     return new PluginClass(BatchSource.PLUGIN_TYPE, "Mock", "", MockSource.class.getName(), "config", properties);
   }
+
+  /**
+   * Processes metadata operations
+   */
+  private void processsMetadata(BatchSourceContext context) throws MetadataException {
+    MetadataEntity metadataEntity = MetadataEntity.ofDataset(context.getNamespace(), config.tableName);
+    Map<MetadataScope, Metadata> curMetadata =
+      context.getMetadata(metadataEntity);
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: the newline is unnecessary/

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java [hunk 4]
```diff
@@ -175,6 +214,52 @@ private static PluginClass getPluginClass() {
...
+    Set<MetadataOperation> operations = GSON.fromJson(config.metadataOperations, SET_METADATA_OPERATION_TYPE);
+    // must be to fetch metadata and there should be system metadata
+    if (curMetadata.get(MetadataScope.SYSTEM).getProperties().isEmpty() ||
+      curMetadata.get(MetadataScope.SYSTEM).getProperties().isEmpty()) {
+      throw new RuntimeException(String.format("System properties or tags for '%s' is empty. " +
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Shouldn't it be `IllegalArgumentException`?

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java [hunk 5]
```diff
@@ -175,6 +214,52 @@ private static PluginClass getPluginClass() {
...
+    properties.put("metadataOperations", new PluginPropertyField("metadataOperations", "", "string", false, false));
     return new PluginClass(BatchSource.PLUGIN_TYPE, "Mock", "", MockSource.class.getName(), "config", properties);
   }
+
+  /**
+   * Processes metadata operations
+   */
+  private void processsMetadata(BatchSourceContext context) throws MetadataException {
+    MetadataEntity metadataEntity = MetadataEntity.ofDataset(context.getNamespace(), config.tableName);
+    Map<MetadataScope, Metadata> curMetadata =
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Call it `currentMetadata`.

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Done.

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java (hunk 3 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Done.

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Done.

File: cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java (hunk 5 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/MockSource.java]: Done.

File: cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-api-spark-base/src/main/java/co/cask/cdap/api/spark/JavaSparkExecutionContextBase.java]: I have exposed them in SparkExecutionContextBase in earlier PR.

**Comment by rohitsinha54** [comment]
@chtyim Addressed comments. Please have another look.

**Comment by rohitsinha54** [comment]
Got some in-person feedback from @chtyim about not exposing the metadata reader and writer to spark transform. Working on the changes. Will re-open PR shortly.

**Review by anew — APPROVED** [review]
no comment from me in addition to @chtyim 's

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java [hunk 6]
```diff
@@ -192,75 +192,119 @@ public URL call() {
...
+      }
     });
   }
 
   @Override
   public void addTags(MetadataEntity metadataEntity, String... tags) {
     CALLER.callUnchecked((Callable<Void>) () -> {
-      pipelineRuntime.getMetadataWriter().addTags(metadataEntity, tags);
-      return null;
+      if (pipelineRuntime.getMetadataWriter().isPresent()) {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java]: Better do

```java
pipelineRuntime.getMetadataWriter()
  .orElseThrow(this::createMetadataUnsupportedException)
  .addTags(metadataEntity, tags);
...

private UnsupportedOperationException createMetadataUnsupportedException() {
  return new UnsupportedOperationException("Metadata operation is not supported.");
}
```

**Comment by rohitsinha54** [comment]
https://builds.cask.co/browse/CDAP-DUT6529-9

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java (hunk 6 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java]: Saw this comment now. Changing.

**Comment by rohitsinha54** [comment]
@chtyim : Please have another look. Thanks.

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java [hunk 7]
```diff
@@ -24,6 +24,8 @@
 import co.cask.cdap.api.data.DatasetContext;
 import co.cask.cdap.api.data.batch.InputContext;
 import co.cask.cdap.api.data.batch.PartitionedFileSetInputContext;
+import co.cask.cdap.api.metadata.MetadataReader;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java]: unused import?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java [hunk 8]
```diff
@@ -192,51 +192,99 @@ public URL call() {
...
+        .addTags(metadataEntity, tags);
+      return null;
+    });
   }
 
   @Override
   public void addTags(MetadataEntity metadataEntity, Iterable<String> tags) {
-
+    CALLER.callUnchecked((Callable<Void>) () -> {
+      pipelineRuntime.getMetadataWriter()
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractStageContext.java]: Just noticed how repetitive it is. Maybe better to have private methods

`private MetadataReader getMetadataReader()` and `private MetadataWriter getMetadataWriter()`

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRuntime.java [hunk 9]
```diff
@@ -103,4 +118,18 @@ public PluginContext getPluginContext() {
   public ServiceDiscoverer getServiceDiscoverer() {
     return serviceDiscoverer;
   }
+
+  public Optional<MetadataReader> getMetadataReader() {
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRuntime.java]: Javadoc for public methods

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java [hunk 10]
```diff
@@ -61,6 +60,8 @@
   private final StageSpec stageSpec;
   private final StageStatisticsCollector collector;
   private transient PipelinePluginContext pipelinePluginContext;
+  private final MetadataReader metadataReader;
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java]: Group `final` fields together.

**Review by chtyim — COMMENTED** [review]
Just couple minor comments.

File: cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java (hunk 7 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceTaskContext.java]: Done.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRuntime.java (hunk 9 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRuntime.java]: Done.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java (hunk 10 above)
Comment (rohitsinha54) [reply cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java]: Done.

**Review by chtyim — APPROVED** [review]
LGTM

**Comment by rohitsinha54** [comment]
Waiting on build..........................

**Comment by prinam** [comment]
Latest build running: https://builds.cask.co/browse/CDAP-DUT6529-18

**Comment by rohitsinha54** [comment]
Waiting on build now running against release/5.0 
https://builds.cask.co/browse/CDAP-RUT1507-1

**Comment by rohitsinha54** [comment]
Build passed merging: https://builds.cask.co/browse/CDAP-RUT1507-6

**Pull request merged**